### PR TITLE
Fix scan failures on nested compiler directives

### LINKED ISSUE #260: Expression lexer failing to compile file in recursive directives
### Prerequisites

- [X] This bug is in SonarDelphi, not SonarQube or my Delphi code.
- [X] This bug has not already been reported.

### SonarDelphi version

1.6.0

### SonarQube version

_No response_

### Issue description

An exception is raised in some delphi files while analyzing it, _au.integradev.delphi.preprocessor.directive.CompilerDirectiveParserImpl$CompilerDirectiveParserError: UNexpected character: '{'_ 

(CompilerDirectiveParserImpl.java) EXPRESSION_LEXER seems to not expect another directive and fails to compile the file.

image
image

Full exception on file: [exception-log.txt](https://github.com/user-attachments/files/15838485/exception-log.txt)

### Steps to reproduce

- Create directive that calls another directive recursively (like the example code below)

- Execute Sonar Scanner analysis with Sonar Delphi

- An exception should be raised in the file with the directive that calls another directive

### Minimal Delphi code exhibiting the issue

```
{$IF LicenseVersion <> {$I __version.txt}}
  {$Message Fatal 'Version different than the expected'}
{$ENDIF}
```

### PR DESCRIPTION
**Pull request by Cirras**
This PR:
- Promotes `Token.TokenType` to public visibility (incidental QA)
- Improves directive name parsing in cases where the name is not followed by whitespace
  - For example, the following is a valid `IF` directive: `{$ifǣ <> 123}`
- Fixes scan failures around nested compiler directives

Note that this does not add proper **support** or evaluation for nested compiler directives.
That would be quite challenging to accomplish without first doing #261.

Fixes #260.

### CODE REVIEW

**Review by denieu — APPROVED** [review]

**Review by zaneduffield — CHANGES_REQUESTED** [review]
Looking pretty good overall. I've left a couple of questions and nit-picky suggestions.

File: delphi-frontend/src/main/antlr3/au/com/integradev/delphi/antlr/Delphi.g
[the remark this answers is not in the document]
Comment (Cirras) [reply delphi-frontend/src/main/antlr3/au/com/integradev/delphi/antlr/Delphi.g:1]: I did, there's nothing to do but live with the warnings in this case.
The warning is a bit naive, we reference both token types programmatically to manually set the token types when parsing reals and multiline strings.

[the remark this answers is not in the document]
Comment (Cirras) [reply delphi-frontend/src/main/antlr3/au/com/integradev/delphi/antlr/Delphi.g:1]: Decided to add support for it.
Added `CompilerDirectiveParserTest::testCreateNestedIfDirectivesWithConfusingStringExpressions`

File: delphi-frontend/src/test/java/au/com/integradev/delphi/preprocessor/directive/CompilerDirectiveParserTest.java
[the remark this answers is not in the document]
Comment (Cirras) [reply delphi-frontend/src/test/java/au/com/integradev/delphi/preprocessor/directive/CompilerDirectiveParserTest.java:1]: Agreed, added `DelphiPreprocessorTest::testPseudoExpressionsShouldNotThrowException`

File: delphi-frontend/src/main/java/au/com/integradev/delphi/preprocessor/directive/CompilerDirectiveParserImpl.java [hunk 1]
```diff
@@ -321,11 +321,11 @@ private boolean isEndOfDirective(char character) {
     boolean result = false;
 
     if (directiveBracketType == CURLY) {
-      result = (character == '}');
+      result = character == '}' && peekChar() == END_OF_INPUT;
     }
 
     if (directiveBracketType == PAREN) {
-      result = (character == '*' && peekChar() == ')');
+      result = character == '*' && peekChar() == ')' && peekChar(1) == END_OF_INPUT;
```
[the remark this answers is not in the document]
Comment (Cirras) [reply delphi-frontend/src/main/java/au/com/integradev/delphi/preprocessor/directive/CompilerDirectiveParserImpl.java:328]: > I'm not sure why this is necessary. I reverted the change and all the tests passed.

Added the egregiously-named `CompilerDirectiveParserTest::testExpressionSurroundedWithNestedDirectivesShouldNotThrowException`

> As an aside, I'm not sure it's safe to represent `END_OF_INPUT` as `\0` since that's a valid whitespace character in Delphi.

This is a good point about the `END_OF_INPUT` sentinel value, but I'm not too bothered about it since this entire separate "compiler directive parser" should really disappear in the eventual future where a more authentic parsing solution is implemented with #261.

If it's a sticking point, we could look at using the same sentinel value as the ANTLR3 generated lexer (to at least minimize the number of magic character values)

File: delphi-frontend/src/main/java/au/com/integradev/delphi/preprocessor/directive/expression/ExpressionLexer.java [hunk 2]
```diff
@@ -259,6 +261,108 @@ private int lookaheadSingleQuotes(int i) {
...
+            break;
+          }
+        }
+        if (j == end.length()) {
+          return i + j;
+        }
+      }
+
+      switch (character) {
+        case '{':
```
[the remark this answers is not in the document]
Comment (Cirras) [reply delphi-frontend/src/main/java/au/com/integradev/delphi/preprocessor/directive/expression/ExpressionLexer.java:394]: Whoops, oversight - added support for this.

Added relevant test data to `DelphiPreprocessorTest::testExpressionsShouldNotThrowException`

File: delphi-frontend/src/test/resources/au/com/integradev/delphi/preprocessor/Expressions.pas [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  {$endif}
+}
+{$endif}
+
+{$if
+  {$if Bracket = '}'}
+  True
+  {$endif}
+}
+{$endif}
```
Comment (zaneduffield) [inline delphi-frontend/src/test/resources/au/com/integradev/delphi/preprocessor/Expressions.pas:48]: This is a duplicate of the previous example.

**Review by zaneduffield — COMMENTED** [review]
Looks like you're handling all those pesky edge cases now. Everything looks good, although I think you duplicated a test case in one of the resources.

File: delphi-frontend/src/main/java/au/com/integradev/delphi/preprocessor/directive/CompilerDirectiveParserImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (zaneduffield) [reply delphi-frontend/src/main/java/au/com/integradev/delphi/preprocessor/directive/CompilerDirectiveParserImpl.java:328]: Honestly, anyone using a null byte as a whitespace character is asking for trouble. It's not a sticking point for me.

File: delphi-frontend/src/test/resources/au/com/integradev/delphi/preprocessor/Expressions.pas (hunk 3 above)
Comment (Cirras) [reply delphi-frontend/src/test/resources/au/com/integradev/delphi/preprocessor/Expressions.pas:48]: Oops, I probably meant to do a variant with `(*$if ... *)`

Comment (Cirras) [reply delphi-frontend/src/test/resources/au/com/integradev/delphi/preprocessor/Expressions.pas]: Added extra test cases for the `(*` bracket style.

**Review by zaneduffield — APPROVED** [review]
Looks good!
